packages/engine/src/worker/runner/python/runner.py

```diff
@@ -1,68 +1,281 @@
+            logging.error("Messenger init failed: %s", error)
+            raise error
 
         self.batches = Batches()
         self.sims = {}
         self.pkgs = {}
+        self.experiment_ctx = None
 
         try:
-            self.experiment_ctx = self.start_experiment()
-        except Exception as e:  # Have to catch generic Exception
-            self.handle_runner_error(e, sys.exc_info())
-            raise e
+            if self.start_experiment():  # Package/user error
+                self._kill_after_sent()
+        except Exception as error:
+            # Have to catch generic Exception -- if we knew what the error
+            # in the runner was, we could have fixed it in the first place.
+            self._handle_runner_error(sys.exc_info())
+            raise error
 
     def start_experiment(self):
+        """
+        Wait for an init message from the Rust process, then use it
+        to initialize experiment-level context (`self.experiment_ctx`)
+        and each package's experiment-level data (`self.pkgs`).
+        The packages' source code is not included in the message,
+        but each package's name and type is included, and its
+        source code is looked up from a path determined by its
+        name and type.
+
+        If an error occurs during a package's custom experiment-level
+        initialization (whether a package or user error), it is sent
+        to the Rust process and experiment init is stopped early and
+        `True` is returned. (`True` is returned instead of raising
+        an exception in order to distinguish runner errors from
+        package/user errors.)
+
+        User warnings are also sent to the Rust process, but do not
+        stop experiment-level init.
+
+        This function is actually only called once per Python process,
+        soon after the process starts, because the init message is
+        always the first message that the Rust process sends to a
+        Python process.
+
+        :return: Whether a package/user error occurred
+        """
         init = self.messenger.recv_init()
-        experiment_ctx = init.shared_ctx
+        self.experiment_ctx = init.shared_ctx
         for pkg_id, config in init.pkgs.items():
             self.pkgs[pkg_id] = pkg = Package(
                 name=config.name,
                 pkg_type=config.type,
-                owned_fields=config.owned_fields
+                owned_fields=[],  # TODO: Propagate `config.owned_fields` here.
             )
 
             if pkg.start_experiment is not None:
-                pkg.start_experiment(
-                    pkg.experiment, config.payload, experiment_ctx
-                )
-
-        return experiment_ctx
+                try:
+                    result = pkg.start_experiment(
+                        pkg.experiment, config.payload, self.experiment_ctx
+                    )
+                    were_user_errors = self._handle_experiment_init_result(pkg, result)
```

Comment:
I think this would also be cleaner if we just propagated a `UserError` exception type up